AWS - Terraform plan with directory services causes panic / crash with segmentation violation code=0x1 addr=0x18 pc=0x40e2395

_This issue was originally opened by @screwnet as hashicorp/terraform#27538. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._
    
<hr>
    
<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers.

To fix problems, we need clear reproduction cases - we need to be able to see it happen locally. A reproduction case is ideally something a Terraform Core engineer can git-clone or copy-paste and run immediately, without inventing any details or context. 

* A short example can be directly copy-pasteable; longer examples should be in separate git repositories, especially if multiple files are needed
* Please include all needed context. For example, if you figured out that an expression can cause a crash, put the expression in a variable definition or a resource
* Set defaults on (or omit) any variables. The person reproducing it should not need to invent variable settings
* If multiple steps are required, such as running terraform twice, consider scripting it in a simple shell script. For example, see [this case](https://github.com/danieldreier/terraform-issue-reproductions/tree/master/25719). Providing a script can be easier than explaining what changes to make to the config between runs.
* Omit any unneeded complexity: remove variables, conditional statements, functions, modules, providers, and resources that are not needed to trigger the bug
* When possible, use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies. We know this isn't always feasible. The Terraform Core team doesn't have deep domain knowledge in every provider, or access to every cloud platform for reproduction cases.

-->

### Terraform Version
<!---
Run `terraform version` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.14.4
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For Terraform configs larger than a few resources, or that involve multiple files, please make a GitHub repository that we can clone, rather than copy-pasting multiple files in here. For security, you can also encrypt the files using our GPG public key at https://www.hashicorp.com/security.
-->
main.tf
```terraform
terraform {

  required_version = "~> 0.14.3"
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.24"
    }
  }
}

provider "aws" {
  region = "us-east-1"
}

data "aws_directory_service_directory" "aws-ad" {
  directory_id = "d-10_char_alphanumeric_id"
}

```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->
```
https://gist.github.com/screwnet/b856bba058839d1d415ec3deb975a1bc
```

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->
```
https://gist.github.com/screwnet/c845f79b170f0befa8a02a578edc0fed
```

### Expected Behavior
<!--

-->
The code should have executed without crashing during terraform plan operation

### Actual Behavior
<!--

-->
When:
```
terraform plan
```
is executed, the code causes Terraform to crash with panic.
### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
Copy the terraform code above to a test file and run following commands in the same directory:
```
terraform init
terraform validate
terraform plan
```

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
Any invalid values in the resource properties were inserted as a part of cleaning logs.
The AD directory is shared from a different AWS account (via VPC peering and handshake) and was not created in this account. The executed users had admin privileges on DS, EC2 services in the destination account where the code was run.
All routing, security group and peering is working fine as well as aws cli:
```
$aws ds describe-directories --directory-ids "d-10_char_alphanumeric_id"

{
    "DirectoryDescriptions": [
        {
            "DirectoryId": "d-10_char_alphanumeric_id",
            "Name": "example.com",
            "ShortName": "EXAMPLE",
            "Size": "Small",
            "Edition": "Standard",
            "Alias": "d-10_char_alphanumeric_id",
            "AccessUrl": "d-10_char_alphanumeric_id.awsapps.com",
            "Stage": "Active",
            "ShareStatus": "Shared",
            "ShareMethod": "HANDSHAKE",
            "LaunchTime": 1606833348.919,
            "StageLastUpdatedDateTime": 1608647353.947,
            "Type": "SharedMicrosoftAD",
            "SsoEnabled": false,
            "DesiredNumberOfDomainControllers": 0,
            "OwnerDirectoryDescription": {
                "DirectoryId": "d-actual_AD_10_char_alphanumeric_id",
                "AccountId": "actual_AD_hosted_AWS_Account_ID",
                "DnsIpAddrs":["192.168.1.1","192.168.2.1"],
                "VpcSettings": {
                    "VpcId": "vpc-actual_AD_hosted_AWS_VPC_IDe",
                    "SubnetIds": [["subnet-sn-id","subnet-sn-id2"],
                    "AvailabilityZones": [
                        "us-east-1a",
                        "us-east-1b"
                    ]
                }
            }
        }
    ]
}

```
### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- hashicorp/terraform#6017
Unaware of
-->

I am having a similar issue while using the data source for `aws_directory_service_directory` that is shared by other aws account. (I am using the shared Directory ID / not the owner's Directory ID)

the setup was tried with below configurations -
**provider versions** - `3.28.0` / `3.27.0`
**Terraform Versions** - `0.12.29` and `0.12.30`

`data "aws_directory_service_directory" "my_ads" {
  directory_id = "d-0000000"
}`

**CRASH LOG**

```
2021-02-19T12:24:40.191+0100 [INFO]  plugin.terraform-provider-aws_v3.28.0_x5: 2021/02/19 12:24:40 [DEBUG] Received DS directory: {
  AccessUrl: "d-000000.awsapps.com",
  Alias: "d-000000",
  DesiredNumberOfDomainControllers: 0,
  DirectoryId: "d-000000",
  Edition: "Standard",
  LaunchTime: 2021-02-18 22:07:18.422 +0000 UTC,
  Name: "my.xxx.local",
  OwnerDirectoryDescription: {
    AccountId: "xxxxxxxx",
    DirectoryId: "d-000000",
    DnsIpAddrs: ["xx.xx.xx.xx","xx.xx.xx.xx"],
    VpcSettings: {
      AvailabilityZones: ["eu-central-1a","eu-central-1b"],
      SubnetIds: ["subnet-xxxxx","subnet-xxxxx"],
      VpcId: "vpc-xxxxx"
    }
  },
  ShareMethod: "HANDSHAKE",
  ShareStatus: "Shared",
  ShortName: "my-ads",
  Size: "Small",
  SsoEnabled: false,
  Stage: "Active",
  StageLastUpdatedDateTime: 2021-02-19 08:57:42.949 +0000 UTC,
  Type: "SharedMicrosoftAD"
}: timestamp=2021-02-19T12:24:40.191+0100
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: panic: runtime error: invalid memory address or nil pointer dereference
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: [signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x4d63075]
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: goroutine 1172 [running]:
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: github.com/terraform-providers/terraform-provider-aws/aws.dataSourceAwsDirectoryServiceDirectoryRead(0xc000db7b80, 0x674f300, 0xc001ed6000, 0xb33adb0, 0x75801a0)
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-aws/aws/data_source_aws_directory_service_directory.go:188 +0xdb5
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc00017fc80, 0x7eb2260, 0xc001b1ca40, 0xc000db7b80, 0x674f300, 0xc001ed6000, 0x0, 0x0, 0x0)
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.3/helper/schema/resource.go:290 +0x88
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).ReadDataApply(0xc00017fc80, 0x7eb2260, 0xc001b1ca40, 0xc0017ed600, 0x674f300, 0xc001ed6000, 0xc001ed6000, 0xc0017ed600, 0x0, 0x0)
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.3/helper/schema/resource.go:498 +0xfd
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadDataSource(0xc00092e440, 0x7eb2260, 0xc001b1ca40, 0xc0017ed340, 0xc001b1ca40, 0x6fa60a0, 0x7245420)
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.3/helper/schema/grpc_provider.go:1105 +0x4d4
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ReadDataSource(0xc00189d420, 0x7eb2260, 0xc001b1ca40, 0xc001d378b0, 0xc00189d420, 0xc001a5fbf0, 0xc001604ba0)
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.2.1/tfprotov5/server/server.go:247 +0xe5
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler(0x7245420, 0xc00189d420, 0x7eb2320, 0xc001a5fbf0, 0xc000c50ae0, 0x0, 0x7eb2320, 0xc001a5fbf0, 0xc0018c6e10, 0xe6)
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.2.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:416 +0x214
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001ff880, 0x7ed57c0, 0xc0009b4300, 0xc001415200, 0xc000ddac60, 0xb2f9ad0, 0x0, 0x0, 0x0)
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1194 +0x522
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: google.golang.org/grpc.(*Server).handleStream(0xc0001ff880, 0x7ed57c0, 0xc0009b4300, 0xc001415200, 0x0)
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1517 +0xd05
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000f6e000, 0xc0001ff880, 0x7ed57c0, 0xc0009b4300, 0xc001415200)
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:859 +0xa5
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: created by google.golang.org/grpc.(*Server).serveStreams.func1
2021-02-19T12:24:40.194+0100 [DEBUG] plugin.terraform-provider-aws_v3.28.0_x5: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd
2021/02/19 12:24:40 [ERROR] module.deliver_event_loggroup: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2021/02/19 12:24:40 [ERROR] <root>: eval: *terraform.EvalReadData, err: rpc error: code = Unavailable desc = transport is closing
2021/02/19 12:24:40 [ERROR] <root>: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2021/02/19 12:24:40 [TRACE] [walkRefresh] Exiting eval tree: data.aws_directory_service_directory.bais_ads
2021/02/19 12:24:40 [TRACE] vertex "data.aws_directory_service_directory.my_ads": visit complete
2021-02-19T12:24:40.203+0100 [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-02-19T12:24:40.203+0100 [DEBUG] plugin: plugin process exited: path=/myfolder/.terraform/plugins/darwin_amd64/terraform-provider-aws_v3.28.0_x5 pid=94502 error="exit status 2"
2021/02/19 12:24:40 [TRACE] <root>: eval: *terraform.EvalSequence
2021/02/19 12:24:40 [TRACE] <root>: eval: *terraform.EvalGetProvider
```

--- a/.changelog/19820.txt
+++ b/.changelog/19820.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+data-source/aws_directory_service_directory: Check VpcSettings and ConnectSettings for nil values
+```

--- a/aws/data_source_aws_directory_service_directory.go
+++ b/aws/data_source_aws_directory_service_directory.go
@@ -182,9 +182,9 @@ func dataSourceAwsDirectoryServiceDirectoryRead(d *schema.ResourceData, meta int
 	d.Set("enable_sso", dir.SsoEnabled)
 
 	var securityGroupId *string
-	if aws.StringValue(dir.Type) == directoryservice.DirectoryTypeAdconnector {
+	if aws.StringValue(dir.Type) == directoryservice.DirectoryTypeAdconnector && dir.ConnectSettings != nil {
 		securityGroupId = dir.ConnectSettings.SecurityGroupId
-	} else {
+	} else if dir.VpcSettings != nil {
 		securityGroupId = dir.VpcSettings.SecurityGroupId
 	}
 	d.Set("security_group_id", securityGroupId)

